File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        HashSet set = new HashSet<>();
        ListNode curr = head;
        while (curr != null) {
            if (set.contains(curr)){
                return true;
            } else {
                set.add(curr);
                curr = curr.next;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        # need to set new variables to head otherwise we lose rest of the list
        prev = None
        curr = head
        
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev # bc curr is now None

    # we didnt make a new copy of the list, just made new pointers/references to the nodes